First commits

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR includes the initial functionality of this project, which is to create an IAM group containing policies that allow members of the group to assume the IAM roles necessary to manage (create, update, delete) IAM users and groups in the COOL.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This will enable the COOL development team to hand production support of COOL IAM user management over to the Operations support team.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I applied this Terraform in Production and verified that a member of the Operations support team was able to:

- [x] Add a new COOL IAM user
- [x] Enable console access for a COOL IAM user
- [x] Add a COOL IAM user to a group
- [x] Remove a COOL IAM user from a group
- [x] Delete a virtual MFA device from a COOL IAM user 
- [x] Disable console access for a COOL IAM user
- [x] Delete an existing COOL IAM user

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

dav3r committed:
Remove unneeded examples directory

dav3r committed:
Remove unused file inherited from skeleton

dav3r committed:
Define S3 backend

dav3r committed:
Specify remote TF states

dav3r committed:
Define variables

dav3r committed:
Define providers

dav3r committed:
Set up useful local variables

dav3r committed:
Define policy containing roles IAM managers need to assume

dav3r committed:
Create IAM managers group and populate it

dav3r committed:
Replace skeleton outputs with real outputs

dav3r committed:
Update README from skeleton with useful information

dav3r committed:
Add an additional policy and rename an existing one

Also, update the outputs with the new policy and updated name.

dav3r committed:
Add new policy to IAM managers group

Also, handle renamed policy.

dav3r committed:
Update Terraform documentation to reflect new and updated policies

dav3r committed:
Add iam:ListGroupsForUser permission to policy

This will allow the IAM manager to confirm  which groups a user belongs to.

mcdonnnj commented on outputs.tf:
```diff
@@ -1,24 +1,14 @@
-output "arn" {
-  value       = aws_instance.example.arn
-  description = "The EC2 instance ARN."
+output "iam_managers_group" {
```
I think this output should be last for the output names to be sorted alphabetically.

dav3r replied:
Good catch - corrected with https://github.com/cisagov/cool-iam-user-group-manager-iam/pull/4/commits/1ef77e154c326a556ebb4b6e7ac5d86a33a05a85.

mcdonnnj commented on locals.tf:
```diff
@@ -0,0 +1,27 @@
+
+  # Assumption of the following roles is required to successfully manage IAM
+  # users and groups.
+  required_roles = [
```
I may be overthinking things but this list of roles seems to allow very broad permissions for a group that is only intended to manage users and groups.

dav3r replied:
The reason this is like this is because the repos that we use to manage users and groups rely on assumption of these roles.  For reference, here are the providers in the repos that use those roles:

- https://github.com/cisagov/cool-users-non-admin/blob/develop/providers.tf
- https://github.com/cisagov/cool-admin-provisioner-iam/blob/develop/providers.tf
- https://github.com/cisagov/cool-assessment-provisioner-iam/blob/develop/providers.tf
- https://github.com/cisagov/cool-auditor-iam/blob/develop/providers.tf
- https://github.com/cisagov/cool-ses-send-email-iam/blob/develop/providers.tf

If we were to modify those repos to use more specific (i.e. less powerful) roles, then we could change the roles here.

For now, how about I add a comment that explains all of this a bit more and provides links to those providers files?

mcdonnnj replied:
Alright. My main concern is that [cisagov/provisionaccount-role-tf-module](https://github.com/cisagov/provisionaccount-role-tf-module) includes the `IAMFullAccess` managed policy which, per the [JSON document](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/IAMFullAccess.html#IAMFullAccess-json), includes the `iam:*` permission.

dav3r replied:
Let me know how https://github.com/cisagov/cool-iam-user-group-manager-iam/pull/4/commits/c3fabef279d4ab336ad164968b0604f3ffb1e4fa works for you.

dav3r replied:
@mcdonnnj Per our chat earlier, please see https://github.com/cisagov/cool-system-internal/issues/137 and https://github.com/cisagov/cool-iam-user-group-manager-iam/pull/4/commits/5d44fbb89adde7ec74283436631221120c937d7c.  Let me know if that is sufficient to resolve this conversation.

mcdonnnj replied:
Yes that works for me. Thanks for creating that issue and updating the comment 🏆

mcdonnnj commented on policies.tf:
```diff
@@ -0,0 +1,70 @@
+
+    resources = local.required_roles
+
+    sid = "AllowRequiredRoles"
```
Something like this seems more in-line with this document.

```suggestion
    sid = "AssumeRequiredRoles"
```

mcdonnnj commented on variables.tf:
```diff
@@ -4,30 +4,55 @@
-  description = "The ID of the AWS subnet to deploy into (e.g. subnet-0123456789abcdef0)."
+variable "users" {
+  type        = list(string)
+  description = "A list containing the usernames of users that exist in the Users account who are allowed to manage IAM users and groups.  Example: [ \"firstname1.lastname1\", \"firstname2.lastname2\" ]."
```
Consistent with the other variables.

```suggestion
  description = "A list containing the usernames of users that exist in the Users account who are allowed to manage IAM users and groups.  Example: [ \"firstname1.lastname1\", \"firstname2.lastname2\" ]."
  type        = list(string)
```

mcdonnnj reviewed (changes requested):
Strong work 💪. I have some minor suggestions and a reservation for your consideration.

dav3r committed:
Sort outputs alphabetically

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Maintain alphabetical ordering of variable keys

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve policy statement ID string

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve a comment to explain source of roles list

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Reference a related issue in a comment

Use more targeted permissions to manage user roles: https://github.com/cisagov/cool-system-internal/issues/137

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/audit.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/audit.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/dns.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/dns.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/images.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/images.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/logarchive.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/logarchive.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/master.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/master.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/shared_services.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/shared_services.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/terraform.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/terraform.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on remote_states.tf:
```diff
@@ -0,0 +1,137 @@
+  backend = "s3"
+
+  config = {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-accounts/users.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-accounts/users.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj commented on outputs.tf:
```diff
@@ -1,24 +1,14 @@
-  value       = aws_instance.example.arn
-  description = "The EC2 instance ARN."
+output "iam_manager_login_mfa_policy" {
+  value       = aws_iam_policy.iam_manager_login_mfa
+  description = "The IAM policy in the Users account that allows the IAM managers group to manage login profiles and MFA devices for IAM users."
```
```suggestion
  description = "The IAM policy in the Users account that allows the IAM managers group to manage login profiles and MFA devices for IAM users."
  value       = aws_iam_policy.iam_manager_login_mfa
```

mcdonnnj commented on outputs.tf:
```diff
@@ -1,24 +1,14 @@
-  value       = aws_instance.example.availability_zone
-  description = "The AZ where the EC2 instance is deployed."
+output "iam_manager_roles_policy" {
+  value       = aws_iam_policy.iam_manager_roles
+  description = "The IAM policy in the Users account that allows the IAM managers group to assume all roles needed in order to manage IAM users and groups."
```
```suggestion
  description = "The IAM policy in the Users account that allows the IAM managers group to assume all roles needed in order to manage IAM users and groups."
  value       = aws_iam_policy.iam_manager_roles
```

mcdonnnj commented on outputs.tf:
```diff
@@ -1,24 +1,14 @@
-  value       = aws_instance.example.subnet_id
-  description = "The ID of the subnet where the EC2 instance is deployed."
+output "iam_managers_group" {
+  value       = aws_iam_group.iam_user_group_managers
+  description = "The IAM group whose members are allowed to manage IAM users and groups."
```
```suggestion
  description = "The IAM group whose members are allowed to manage IAM users and groups."
  value       = aws_iam_group.iam_user_group_managers
```

mcdonnnj commented on group_membership.tf:
```diff
@@ -0,0 +1,11 @@
+  provider = aws.users
+  for_each = toset(var.users)
+
+  user = each.key
+
+  groups = [
+    aws_iam_group.iam_user_group_managers.name
+  ]
```
```suggestion
  groups = [
    aws_iam_group.iam_user_group_managers.name
  ]

  user = each.key
```

mcdonnnj commented on backend.tf:
```diff
@@ -0,0 +1,10 @@
+terraform {
+  backend "s3" {
+    encrypt        = true
+    bucket         = "cisa-cool-terraform-state"
+    dynamodb_table = "terraform-state-lock"
+    profile        = "cool-terraform-backend"
+    region         = "us-east-1"
+    key            = "cool-iam-user-group-manager-iam/terraform.tfstate"
```
```suggestion
    bucket         = "cisa-cool-terraform-state"
    dynamodb_table = "terraform-state-lock"
    encrypt        = true
    key            = "cool-iam-user-group-manager-iam/terraform.tfstate"
    profile        = "cool-terraform-backend"
    region         = "us-east-1"
```

mcdonnnj reviewed (commented):
Alphabetization recommendations just to be consistent since you've already done it to the variables.

dav3r committed:
Alphabetize a bunch of thangs

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
